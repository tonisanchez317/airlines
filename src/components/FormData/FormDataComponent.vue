<script>
import COUNTRIES from './data/countries.json';

export default {
  name: 'FormData',

  props: {
    value: {
      type: Object,
      default: null,
    },
  },

  computed: {
    state() {
      return this.value ? this.value : {};
    },
  },

  created() {
    this.countries = COUNTRIES.data;
  },
};
</script>
<template>
  <form>
    <div
      v-if="state.id"
      class="form-group"
    >
      <label>
        ID
      </label>
      <div>
        {{ state.id }}
      </div>
    </div>

    <div class="form-group">
      <label for="iata">
        IATA
      </label>
      <input
        id="iata"
        type="text"
        class="form-control"
        placeholder="IATA"
        v-model="state.iata"
      >
    </div>

    <div class="form-group">
      <label for="icao">
        ICAO
      </label>
      <input
        id="icao"
        type="text"
        class="form-control"
        placeholder="ICAO"
        v-model="state.icao"
      >
    </div>

    <div class="form-group">
      <label for="airline">
        Airline
      </label>
      <input
        id="airline"
        type="text"
        class="form-control"
        placeholder="Airline"
        v-model="state.airline"
      >
    </div>

    <div class="form-group">
      <label for="callsign">
        Call sign
      </label>
      <input
        id="callsign"
        type="text"
        class="form-control"
        placeholder="Call sign"
        v-model="state.callsign"
      >
    </div>

    <div class="form-group">
      <label for="country">
        Country
      </label>
      <select
        id="country"
        class="form-control"
        v-model="state.country"
      >
        <option
          :value="undefined"
        >
          Choose a country
        </option>
        <option
          v-for="(country, countryIndex) in countries"
          :key="`country-${countryIndex}`"
          :value="country"
        >
          {{ country }}
        </option>
      </select>
    </div>
  </form>
</template>
