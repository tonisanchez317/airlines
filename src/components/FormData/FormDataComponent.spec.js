import { shallowMount, mount } from '@vue/test-utils';
import FormDataComponent from './FormDataComponent.vue';

let wrapper;

jest.mock('./data/countries.json', () => ({
  data: [
    'Spain',
    'USA',
    'France',
  ],
}));

describe('Given FormDataComponent', () => {
  describe('When component does not receive a value', () => {
    beforeEach(() => {
      wrapper = shallowMount(FormDataComponent);
    });
    it('Then state computed should be an empty object', () => {
      expect(wrapper.vm.value).toEqual(null);
    });
    it('Then HTML rendered should be the same than original snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });
  describe('When component receives a value', () => {
    beforeEach(() => {
      wrapper = mount(FormDataComponent, {
        propsData: {
          value: {
            id: 1,
            iata: 'iata test',
            icao: 'icao test',
            airline: 'airline test',
            callsign: 'callsign test',
            country: 'Spain',
          },
        },
      });
    });
    describe('And this data is edited by the user', () => {
      it('Then value should be updated', async () => {
        const iata = wrapper.find('#iata');
        await iata.setValue('new iata value');

        const icao = wrapper.find('#icao');
        await icao.setValue('new icao value');

        const airline = wrapper.find('#airline');
        await airline.setValue('new airline value');

        const callsign = wrapper.find('#callsign');
        await callsign.setValue('new callsign value');

        const country = wrapper.find('#country');
        await country.setValue('USA');

        expect(wrapper.vm.value.iata).toEqual('new iata value');
        expect(wrapper.vm.value.icao).toEqual('new icao value');
        expect(wrapper.vm.value.airline).toEqual('new airline value');
        expect(wrapper.vm.value.callsign).toEqual('new callsign value');
        expect(wrapper.vm.value.country).toEqual('USA');
      });
    });
    it('Then value prop should be returned as state', () => {
      expect(wrapper.vm.state).toEqual({
        id: 1,
        iata: 'iata test',
        icao: 'icao test',
        airline: 'airline test',
        callsign: 'callsign test',
        country: 'Spain',
      });
    });
    it('Then HTML rendered should be the same than original snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });
    it('Then inputs should be filled with that data', () => {
      expect(wrapper.find('#iata').element.value).toEqual('iata test');
      expect(wrapper.find('#icao').element.value).toEqual('icao test');
      expect(wrapper.find('#airline').element.value).toEqual('airline test');
      expect(wrapper.find('#callsign').element.value).toEqual('callsign test');
      expect(wrapper.find('#country').element.value).toEqual('Spain');
    });
  });
});
