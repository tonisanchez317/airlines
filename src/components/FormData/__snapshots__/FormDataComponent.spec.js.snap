// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given FormDataComponent When component does not receive a value Then HTML rendered should be the same than original snapshot 1`] = `
<form>
  <!---->
  <div class="form-group"><label for="iata">
      IATA
    </label> <input id="iata" type="text" placeholder="IATA" class="form-control"></div>
  <div class="form-group"><label for="icao">
      ICAO
    </label> <input id="icao" type="text" placeholder="ICAO" class="form-control"></div>
  <div class="form-group"><label for="airline">
      Airline
    </label> <input id="airline" type="text" placeholder="Airline" class="form-control"></div>
  <div class="form-group"><label for="callsign">
      Call sign
    </label> <input id="callsign" type="text" placeholder="Call sign" class="form-control"></div>
  <div class="form-group"><label for="country">
      Country
    </label> <select id="country" class="form-control">
      <option value="">
        Choose a country
      </option>
      <option value="Spain">
        Spain
      </option>
      <option value="USA">
        USA
      </option>
      <option value="France">
        France
      </option>
    </select></div>
</form>
`;

exports[`Given FormDataComponent When component receives a value Then HTML rendered should be the same than original snapshot 1`] = `
<form>
  <div class="form-group"><label>
      ID
    </label>
    <div>
      1
    </div>
  </div>
  <div class="form-group"><label for="iata">
      IATA
    </label> <input id="iata" type="text" placeholder="IATA" class="form-control"></div>
  <div class="form-group"><label for="icao">
      ICAO
    </label> <input id="icao" type="text" placeholder="ICAO" class="form-control"></div>
  <div class="form-group"><label for="airline">
      Airline
    </label> <input id="airline" type="text" placeholder="Airline" class="form-control"></div>
  <div class="form-group"><label for="callsign">
      Call sign
    </label> <input id="callsign" type="text" placeholder="Call sign" class="form-control"></div>
  <div class="form-group"><label for="country">
      Country
    </label> <select id="country" class="form-control">
      <option value="">
        Choose a country
      </option>
      <option value="Spain">
        Spain
      </option>
      <option value="USA">
        USA
      </option>
      <option value="France">
        France
      </option>
    </select></div>
</form>
`;
